File: backend/internal/store/store.go
package store

import (
	"fmt"
	"sort"
	"sync"
	"time"

	"github.com/pluszero/dental-care-api/internal/models"
)

type Store struct {
	mu           sync.RWMutex
	patients     map[string]models.Patient
	appointments map[string]models.Appointment
	treatments   map[string]models.TreatmentRecord
	organizations map[string]models.Organization
	users        map[string]models.User
	members      map[string]models.TeamMember
	apiKeys      map[string]models.APIKey
	auditLogs    []models.AuditLogEntry
	currentUserID string
	currentOrgID  string
	patientProfiles map[string]models.PatientProfile
	insurance       map[string]models.InsuranceInfo
	medicalHistories []models.MedicalHistory
	allergyRecords   []models.AllergyRecord
	familyMembers    []models.FamilyMember
	questionnaires   []models.Questionnaire
	emergencyContacts []models.EmergencyContact
	visitRecords     []models.VisitRecord
	chairs           []models.Chair
	staffSchedules   []models.StaffSchedule
	reminders        []models.ReminderNotification
	xrayImages       map[string]models.XrayImage
	perioExams       map[string]models.PerioExam
}

func New() *Store {
	s := &Store{
		patients:      make(map[string]models.Patient),
		appointments:  make(map[string]models.Appointment),
		treatments:    make(map[string]models.TreatmentRecord),
		organizations: make(map[string]models.Organization),
		users:         make(map[string]models.User),
		members:       make(map[string]models.TeamMember),
		apiKeys:       make(map[string]models.APIKey),
		patientProfiles: make(map[string]models.PatientProfile),
		insurance:       make(map[string]models.InsuranceInfo),
		xrayImages:      make(map[string]models.XrayImage),
		perioExams:      make(map[string]models.PerioExam),
	}
	s.seed()
	s.seedSaas()
	s.seedClinical()
	s.seedXrays()
	s.seedPerio()
	return s
}

func (s *Store) seed() {
	now := time.Now()
	today := now.Format("2006-01-02")

	patients := []models.Patient{
		{ID: "p1", ChartNo: "10001", Name: "山田 太郎", Kana: "ヤマダ タロウ", BirthDate: "1985-03-12", Phone: "090-1234-5678", Email: "taro@example.com", Allergies: "ペニシリン", Notes: "定期検診希望", LastVisit: "2026-05-10", CreatedAt: now.Add(-365 * 24 * time.Hour)},
		{ID: "p2", ChartNo: "10002", Name: "佐藤 花子", Kana: "サトウ ハナコ", BirthDate: "1992-07-22", Phone: "080-9876-5432", Email: "hanako@example.com", Allergies: "", Notes: "矯正相談済", LastVisit: "2026-05-15", CreatedAt: now.Add(-200 * 24 * time.Hour)},
		{ID: "p3", ChartNo: "10003", Name: "鈴木 一郎", Kana: "スズキ イチロウ", BirthDate: "1978-11-05", Phone: "070-1111-2222", Email: "", Allergies: "ラテックス", Notes: "", LastVisit: "2026-04-28", CreatedAt: now.Add(-90 * 24 * time.Hour)},
		{ID: "p4", ChartNo: "10004", Name: "田中 美咲", Kana: "タナカ ミサキ", BirthDate: "2001-01-18", Phone: "090-3333-4444", Email: "misaki@example.com", Allergies: "", Notes: "学生割引対象", LastVisit: "2026-05-18", CreatedAt: now.Add(-30 * 24 * time.Hour)},
		{ID: "p5", ChartNo: "10005", Name: "伊藤 翔", Kana: "イトウ ショウ", BirthDate: "1990-06-08", Phone: "090-2222-3333", Email: "sho@example.com", Allergies: "", Notes: "", LastVisit: "2026-05-01", CreatedAt: now.Add(-25 * 24 * time.Hour)},
		{ID: "p6", ChartNo: "10006", Name: "渡辺 由美", Kana: "ワタナベ ユミ", BirthDate: "1988-12-20", Phone: "080-4444-5555", Email: "", Allergies: "イブプロフェン", Notes: "", LastVisit: "2026-04-15", CreatedAt: now.Add(-20 * 24 * time.Hour)},
		{ID: "p7", ChartNo: "10007", Name: "中村 大輔", Kana: "ナカムラ ダイスケ", BirthDate: "1975-09-03", Phone: "070-6666-7777", Email: "daisuke@example.com", Allergies: "", Notes: "インプラント相談", LastVisit: "2026-03-22", CreatedAt: now.Add(-15 * 24 * time.Hour)},
		{ID: "p8", ChartNo: "10008", Name: "小林 あゆみ", Kana: "コバヤシ アユミ", BirthDate: "1998-02-14", Phone: "090-8888-9999", Email: "ayumi@example.com", Allergies: "", Notes: "", LastVisit: "2026-05-12", CreatedAt: now.Add(-10 * 24 * time.Hour)},
	}
	for _, p := range patients {
		s.patients[p.ID] = p
	}

	appointments := []models.Appointment{
		{ID: "a1", PatientID: "p1", Chair: 1, StartAt: today + "T09:00:00", EndAt: today + "T09:30:00", Type: "定期検診", Status: "confirmed", Staff: "Dr. 木村", Notes: ""},
		{ID: "a2", PatientID: "p2", Chair: 2, StartAt: today + "T10:00:00", EndAt: today + "T11:00:00", Type: "矯正調整", Status: "confirmed", Staff: "Dr. 木村", Notes: "ワイヤー交換"},
		{ID: "a3", PatientID: "p4", Chair: 1, StartAt: today + "T11:30:00", EndAt: today + "T12:00:00", Type: "クリーニング", Status: "pending", Staff: "衛生士 林", Notes: ""},
		{ID: "a4", PatientID: "p3", Chair: 3, StartAt: today + "T14:00:00", EndAt: today + "T15:00:00", Type: "CR充填", Status: "confirmed", Staff: "Dr. 木村", Notes: "#16"},
	}
	for _, a := range appointments {
		if p, ok := s.patients[a.PatientID]; ok {
			a.Patient = p.Name
		}
		s.appointments[a.ID] = a
	}

	treatments := []models.TreatmentRecord{
		{ID: "t1", PatientID: "p1", VisitDate: "2026-05-10", Tooth: "16", Procedure: "\u30b9\u30b1\u30fc\u30ea\u30f3\u30b0", ProcedureCode: "M011", Diagnosis: "\u6b6f\u77f3\u6c88\u7740", Fee: 3300, Staff: "\u885b\u751f\u58eb \u6797", Status: "completed", Tags: []string{"\u4e88\u9632"},
			Subjective: "\u6b6f\u78da\u304c\u306a\u3064\u304b\u3057\u3044", Objective: "\u6b6f\u77f3\u9644\u7740\u300116\u304b\u9762\u7740\u819c\u708e", Assessment: "\u6b6f\u77f3\u6c88\u7740", Plan: "\u30b9\u30b1\u30fc\u30ea\u30f3\u30b0\u3001\u6b6f\u78da\u6307\u5c0e", ToothChartJSON: `{"selected":["16"],"conditions":{"16":"calculus"}}`},
		{ID: "t2", PatientID: "p2", VisitDate: "2026-05-15", Tooth: "\u5168\u4f53", Procedure: "\u77ef\u6b63\u8abf\u6574", ProcedureCode: "M301", Diagnosis: "\u53e2\u751f", Fee: 8800, Staff: "Dr. \u6728\u6751", Status: "completed", Tags: []string{"\u77ef\u6b63"},
			Subjective: "\u30ef\u30a4\u30e4\u30fc\u304c\u5f53\u305f\u308b", Objective: "\u53e2\u751f\u3001\u30c1\u30a7\u30b9\u30c8\u30a2\u30fc\u30c1\u306e\u5f53\u305f\u308a", Assessment: "\u77ef\u6b63\u4e2d", Plan: "\u30ef\u30a4\u30e4\u30fc\u4ea4\u63db\u7d99\u7d9a", ToothChartJSON: `{"selected":["11","21"],"conditions":{}}`},
		{ID: "t3", PatientID: "p3", VisitDate: "2026-04-28", Tooth: "26", Procedure: "\u6839\u7ba1\u6cbb\u7642", ProcedureCode: "M012", Diagnosis: "\u6b6f\u9ad3\u708e", Fee: 12000, Staff: "Dr. \u6728\u6751", Status: "in_progress", Tags: []string{"\u4fdd\u5b58"},
			Subjective: "\u51b7\u305f\u3044\u3082\u306e\u304c\u6b21\u306e\u306a\u3044", Objective: "26\u306b\u6b6f\u9ad3\u306e\u6fc0\u75db\u3001\u6df7\u6d8c\u306a\u3057", Assessment: "\u6b6f\u9ad3\u708e", Plan: "\u6839\u7ba1\u5145\u586b\u306e\u7d99\u7d9a", ToothChartJSON: `{"selected":["26"],"conditions":{"26":"endo"}}`},
		{ID: "t4", PatientID: "p4", VisitDate: "2026-05-18", Tooth: "\u5168\u4f53", Procedure: "PMTC", ProcedureCode: "M011", Diagnosis: "\u8efd\u5ea6\u6b6f\u5468\u708e", Fee: 5500, Staff: "\u885b\u751f\u58eb \u6797", Status: "completed", Tags: []string{"\u4e88\u9632", "\u6b6f\u5468"},
			Subjective: "\u306f\u3058\u304d\u306e\u8840\u304c\u51fa\u308b", Objective: "BOP+\u3001\u6b6f\u5468\u888b4mm", Assessment: "\u8efd\u5ea6\u6b6f\u5468\u708e", Plan: "PMTC\u3001\u6b6f\u5468\u6307\u5c0e", ToothChartJSON: `{"selected":[],"conditions":{}}`},
	}
	for _, t := range treatments {
		if p, ok := s.patients[t.PatientID]; ok {
			t.PatientName = p.Name
		}
		s.treatments[t.ID] = t
	}
}

func (s *Store) ListPatients() []models.Patient {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]models.Patient, 0, len(s.patients))
	for _, p := range s.patients {
		out = append(out, p)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].ChartNo < out[j].ChartNo })
	return out
}

func (s *Store) PaginatePatients(page, pageSize int) models.PatientPage {
	all := s.ListPatients()
	items, total, p, ps, tp := slicePage(all, page, pageSize)
	return models.PatientPage{
		Items: items,
		PageInfo: models.PageInfo{Total: total, Page: p, PageSize: ps, TotalPages: tp},
	}
}

func (s *Store) GetPatient(id string) (models.Patient, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	p, ok := s.patients[id]
	return p, ok
}

func (s *Store) CreatePatient(p models.Patient) models.Patient {
	s.mu.Lock()
	defer s.mu.Unlock()
	p.CreatedAt = time.Now()
	s.patients[p.ID] = p
	return p
}

func (s *Store) ListAppointments(date string) []models.Appointment {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]models.Appointment, 0)
	for _, a := range s.appointments {
		if date == "" || len(a.StartAt) >= 10 && a.StartAt[:10] == date {
			out = append(out, a)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].StartAt < out[j].StartAt })
	return out
}

func (s *Store) PaginateAppointments(date string, page, pageSize int) models.AppointmentPage {
	all := s.ListAppointments(date)
	items, total, p, ps, tp := slicePage(all, page, pageSize)
	return models.AppointmentPage{
		Items: items,
		PageInfo: models.PageInfo{Total: total, Page: p, PageSize: ps, TotalPages: tp},
	}
}

func (s *Store) CreateAppointment(a models.Appointment) models.Appointment {
	s.mu.Lock()
	defer s.mu.Unlock()
	if p, ok := s.patients[a.PatientID]; ok {
		a.Patient = p.Name
	}
	s.appointments[a.ID] = a
	return a
}

func (s *Store) ListTreatments(patientID string) []models.TreatmentRecord {
	s.mu.RLock()
	defer s.mu.RUnlock()
	out := make([]models.TreatmentRecord, 0)
	for _, t := range s.treatments {
		if patientID == "" || t.PatientID == patientID {
			out = append(out, t)
		}
	}
	sort.Slice(out, func(i, j int) bool { return out[i].VisitDate > out[j].VisitDate })
	return out
}

func (s *Store) PaginateTreatments(patientID string, page, pageSize int) models.TreatmentPage {
	all := s.ListTreatments(patientID)
	items, total, p, ps, tp := slicePage(all, page, pageSize)
	return models.TreatmentPage{
		Items: items,
		PageInfo: models.PageInfo{Total: total, Page: p, PageSize: ps, TotalPages: tp},
	}
}

func (s *Store) Dashboard() models.DashboardStats {
	s.mu.RLock()
	defer s.mu.RUnlock()
	today := time.Now().Format("2006-01-02")
	apptToday := 0
	revenue := 0
	for _, a := range s.appointments {
		if len(a.StartAt) >= 10 && a.StartAt[:10] == today {
			apptToday++
		}
	}
	for _, t := range s.treatments {
		if len(t.VisitDate) >= 7 && t.VisitDate[:7] == time.Now().Format("2006-01") {
			revenue += t.Fee
		}
	}
	noShow := 0
	for _, a := range s.appointments {
		if a.NoShow || a.Status == "no_show" {
			noShow++
		}
	}
	noShowRate := 0.0
	if len(s.appointments) > 0 {
		noShowRate = float64(noShow) / float64(len(s.appointments))
	}
	return models.DashboardStats{
		PatientsTotal:       len(s.patients),
		AppointmentsToday:   apptToday,
		TreatmentsThisMonth: len(s.treatments),
		RevenueThisMonth:    revenue,
		ChairUtilization:    0.72,
		NoShowRate:          noShowRate,
	}
}

func NextID(prefix string, n int) string {
	return fmt.Sprintf("%s%d", prefix, n+1)
}

File: frontend/src/lib/embedded-api/store.ts
import { isoNow, slicePage, todayISO } from './helpers'

type Patient = {
  id: string
  chartNo: string
  name: string
  kana: string
  birthDate: string
  phone: string
  email: string
  allergies: string
  notes: string
  lastVisit: string
  createdAt: string
}

type Appointment = {
  id: string
  patientId: string
  patientName: string
  chair: number
  startAt: string
  endAt: string
  type: string
  status: string
  staff: string
  staffRole: string
  notes: string
  noShow: boolean
  cancelReason?: string | null
  cancelledAt?: string | null
}

type Treatment = {
  id: string
  patientId: string
  patientName: string
  visitDate: string
  tooth: string
  procedure: string
  procedureCode: string
  diagnosis: string
  fee: number
  staff: string
  status: string
  tags: string[]
  subjective: string
  objective: string
  assessment: string
  plan: string
  toothChartJson: string
}

type PerioSite = {
  tooth: string
  pdMesial: number
  pdBuccal: number
  pdDistal: number
  pdLingual: number
  bop: boolean
  mobility: number
}

type PerioExam = {
  id: string
  patientId: string
  examDate: string
  staff: string
  notes: string
  sites: PerioSite[]
  createdAt: string
}

class EmbeddedStore {
  patients: Patient[] = []
  appointments: Appointment[] = []
  treatments: Treatment[] = []
  xrays: Array<{
    id: string
    patientId: string
    title: string
    imageUrl: string
    imageType: string
    toothRegion: string
    takenAt: string
    notes: string
    createdAt: string
    updatedAt: string
  }> = []
  perioExams: PerioExam[] = []

  auditLogs = [
    {
      id: 'log1',
      action: 'member.invited',
      resource: 'team',
      actorName: 'Kimura',
      ipAddress: '203.0.113.10',
      metadata: '{"email":"sato@sakura-dental.jp","role":"MEMBER"}',
      createdAt: isoNow(),
    },
  ]

  constructor() {
    this.seed()
  }

  private seed() {
    const today = todayISO()
    const created = isoNow()

    this.patients = [
      { id: 'p1', chartNo: '10001', name: '\u5c71\u7530 \u592a\u90ce', kana: '\u30e4\u30de\u30c0 \u30bf\u30ed\u30a6', birthDate: '1985-03-12', phone: '090-1234-5678', email: 'taro@example.com', allergies: '\u30da\u30cb\u30b7\u30ea\u30f3', notes: '\u5b9a\u671f\u691c\u8a3a\u5e0c\u671b', lastVisit: '2026-05-10', createdAt: created },
      { id: 'p2', chartNo: '10002', name: '\u4f50\u85e4 \u82b1\u5b50', kana: '\u30b5\u30c8\u30a6 \u30cf\u30ca\u30b3', birthDate: '1992-07-22', phone: '080-9876-5432', email: 'hanako@example.com', allergies: '', notes: '\u77ef\u6b63\u76f8\u8ac7\u6e08', lastVisit: '2026-05-15', createdAt: created },
      { id: 'p3', chartNo: '10003', name: '\u9234\u6728 \u4e00\u90ce', kana: '\u30b9\u30ba\u30ad \u30a4\u30c1\u30ed\u30a6', birthDate: '1978-11-05', phone: '070-1111-2222', email: '', allergies: '\u30e9\u30c6\u30c3\u30af\u30b9', notes: '', lastVisit: '2026-04-28', createdAt: created },
      { id: 'p4', chartNo: '10004', name: '\u7530\u4e2d \u7f8e\u54b2', kana: '\u30bf\u30ca\u30ab \u30df\u30b5\u30ad', birthDate: '2001-01-18', phone: '090-3333-4444', email: 'misaki@example.com', allergies: '', notes: '\u5b66\u751f\u5272\u5f15\u5bfe\u8c61', lastVisit: '2026-05-18', createdAt: created },
      { id: 'p5', chartNo: '10005', name: '\u4f0a\u85e4 \u7fd4', kana: '\u30a4\u30c8\u30a6 \u30b7\u30e7\u30a6', birthDate: '1990-06-08', phone: '090-2222-3333', email: 'sho@example.com', allergies: '', notes: '', lastVisit: '2026-05-01', createdAt: created },
      { id: 'p6', chartNo: '10006', name: '\u6e21\u8fba \u7531\u7f8e', kana: '\u30ef\u30bf\u30ca\u30d9 \u30e6\u30df', birthDate: '1988-12-20', phone: '080-4444-5555', email: '', allergies: '\u30a4\u30d6\u30d7\u30ed\u30d5\u30a7\u30f3', notes: '', lastVisit: '2026-04-15', createdAt: created },
      { id: 'p7', chartNo: '10007', name: '\u4e2d\u6751 \u5927\u8f14', kana: '\u30ca\u30ab\u30e0\u30e9 \u30c0\u30a4\u30b9\u30b1', birthDate: '1975-09-03', phone: '070-6666-7777', email: 'daisuke@example.com', allergies: '', notes: '\u30a4\u30f3\u30d7\u30e9\u30f3\u30c8\u76f8\u8ac7', lastVisit: '2026-03-22', createdAt: created },
      { id: 'p8', chartNo: '10008', name: '\u5c0f\u6797 \u3042\u3086\u307f', kana: '\u30b3\u30d0\u30e4\u30b7 \u30a2\u30e6\u30df', birthDate: '1998-02-14', phone: '090-8888-9999', email: 'ayumi@example.com', allergies: '', notes: '', lastVisit: '2026-05-12', createdAt: created },
    ]

    this.appointments = [
      { id: 'a1', patientId: 'p1', patientName: '\u5c71\u7530 \u592a\u90ce', chair: 1, startAt: `${today}T09:00:00`, endAt: `${today}T09:30:00`, type: '\u5b9a\u671f\u691c\u8a3a', status: 'confirmed', staff: 'Dr. \u6728\u6751', staffRole: 'dentist', notes: '', noShow: false },
      { id: 'a2', patientId: 'p2', patientName: '\u4f50\u85e4 \u82b1\u5b50', chair: 2, startAt: `${today}T10:00:00`, endAt: `${today}T11:00:00`, type: '\u77ef\u6b63\u8abf\u6574', status: 'confirmed', staff: 'Dr. \u6728\u6751', staffRole: 'dentist', notes: '\u30ef\u30a4\u30e4\u30fc\u4ea4\u63db', noShow: false },
      { id: 'a3', patientId: 'p4', patientName: '\u7530\u4e2d \u7f8e\u54b2', chair: 1, startAt: `${today}T11:30:00`, endAt: `${today}T12:00:00`, type: '\u30af\u30ea\u30fc\u30cb\u30f3\u30b0', status: 'pending', staff: '\u885b\u751f\u58eb \u6797', staffRole: 'hygienist', notes: '', noShow: false },
      { id: 'a4', patientId: 'p3', patientName: '\u9234\u6728 \u4e00\u90ce', chair: 3, startAt: `${today}T14:00:00`, endAt: `${today}T15:00:00`, type: 'CR\u5145\u586b', status: 'confirmed', staff: 'Dr. \u6728\u6751', staffRole: 'dentist', notes: '#16', noShow: false },
    ]

    this.treatments = [
      { id: 't1', patientId: 'p1', patientName: '\u5c71\u7530 \u592a\u90ce', visitDate: '2026-05-10', tooth: '16', procedure: '\u30b9\u30b1\u30fc\u30ea\u30f3\u30b0', procedureCode: 'M011', diagnosis: '\u6b6f\u77f3\u6c89\u7740', fee: 3300, staff: '\u885b\u751f\u58eb \u6797', status: 'completed', tags: ['\u4e88\u9632'], subjective: '\u6b6f\u78da\u304c\u306a\u3064\u304b\u3057\u3044', objective: '\u6b6f\u77f3\u9644\u7740', assessment: '\u6b6f\u77f3\u6c89\u7740', plan: '\u30b9\u30b1\u30fc\u30ea\u30f3\u30b0', toothChartJson: '{"selected":["16"],"conditions":{"16":"calculus"}}' },
      { id: 't2', patientId: 'p2', patientName: '\u4f50\u85e4 \u82b1\u5b50', visitDate: '2026-05-15', tooth: '\u5168\u4f53', procedure: '\u77ef\u6b63\u8abf\u6574', procedureCode: 'M301', diagnosis: '\u53c2\u751f', fee: 8800, staff: 'Dr. \u6728\u6751', status: 'completed', tags: ['\u77ef\u6b63'], subjective: '', objective: '', assessment: '', plan: '', toothChartJson: '{"selected":[],"conditions":{}}' },
      { id: 't3', patientId: 'p3', patientName: '\u9234\u6728 \u4e00\u90ce', visitDate: '2026-04-28', tooth: '26', procedure: '\u6839\u7ba1\u6cbb\u7642', procedureCode: 'M012', diagnosis: '\u6b6f\u9ad3\u708e', fee: 12000, staff: 'Dr. \u6728\u6751', status: 'in_progress', tags: ['\u4fdd\u5b58'], subjective: '', objective: '', assessment: '', plan: '', toothChartJson: '{"selected":["26"],"conditions":{"26":"endo"}}' },
      { id: 't4', patientId: 'p4', patientName: '\u7530\u4e2d \u7f8e\u54b2', visitDate: '2026-05-18', tooth: '\u5168\u4f53', procedure: 'PMTC', procedureCode: 'M401', diagnosis: '\u8efd\u5ea6\u6b6f\u5468\u75c5', fee: 5500, staff: '\u885b\u751f\u58eb \u6797', status: 'completed', tags: ['\u4e88\u9632', '\u6b6f\u5468'], subjective: '', objective: '', assessment: '', plan: '', toothChartJson: '{"selected":[],"conditions":{}}' },
    ]

    this.xrays = [
      { id: 'x1', patientId: 'p1', title: 'Panoramic', imageUrl: '/uploads/xrays/placeholder.svg', imageType: 'PANORAMIC', toothRegion: '\u5168\u4f53', takenAt: '2026-05-10', notes: '', createdAt: created, updatedAt: created },
      { id: 'x2', patientId: 'p1', title: '\u53e3\u8154\u5185\u7167', imageUrl: '/uploads/xrays/placeholder.svg', imageType: 'INTRAORAL', toothRegion: '16', takenAt: '2026-05-10', notes: '', createdAt: created, updatedAt: created },
    ]

    this.perioExams = [
      {
        id: 'pe1', patientId: 'p1', examDate: '2026-05-10', staff: '\u885b\u751f\u58eb \u6797', notes: '\u521d\u56de\u691c\u67fb', createdAt: created,
        sites: [{ tooth: '16', pdMesial: 2, pdBuccal: 3, pdDistal: 2, pdLingual: 2, bop: true, mobility: 0 }],
      },
    ]
  }

  dashboard() {
    const today = todayISO()
    const apptToday = this.appointments.filter((a) => a.startAt.startsWith(today)).length
    const revenue = this.treatments.reduce((s, t) => s + t.fee, 0)
    return {
      patientsTotal: this.patients.length,
      appointmentsToday: apptToday,
      treatmentsThisMonth: this.treatments.length,
      revenueThisMonth: revenue,
      chairUtilization: 0.72,
      noShowRate: 0.05,
    }
  }

  paginatePatients(page?: number, pageSize?: number) {
    const sorted = [...this.patients].sort((a, b) => a.chartNo.localeCompare(b.chartNo))
    return slicePage(sorted, page, pageSize)
  }

  patient(id: string) {
    return this.patients.find((p) => p.id === id) ?? null
  }

  paginateAppointments(date?: string, page?: number, pageSize?: number) {
    const today = date || todayISO()
    const filtered = this.appointments
      .filter((a) => a.startAt.slice(0, 10) === today)
      .sort((a, b) => a.startAt.localeCompare(b.startAt))
    return slicePage(filtered, page, pageSize)
  }

  paginateTreatments(patientId?: string, page?: number, pageSize?: number) {
    const filtered = this.treatments.filter((t) => !patientId || t.patientId === patientId)
    return slicePage(filtered, page, pageSize)
  }

  treatment(id: string) {
    return this.treatments.find((t) => t.id === id) ?? null
  }

  createTreatment(input: Omit<Treatment, 'id' | 'patientName'>) {
    const patient = this.patient(input.patientId)
    const record: Treatment = {
      ...input,
      id: `t${Date.now()}`,
      patientName: patient?.name ?? '',
      procedureCode: input.procedureCode ?? '',
      subjective: input.subjective ?? '',
      objective: input.objective ?? '',
      assessment: input.assessment ?? '',
      plan: input.plan ?? '',
      toothChartJson: input.toothChartJson ?? '{"selected":[],"conditions":{}}',
      tags: input.tags ?? [],
    }
    this.treatments.unshift(record)
    return record
  }

  updateTreatment(id: string, patch: Partial<Treatment>) {
    const idx = this.treatments.findIndex((t) => t.id === id)
    if (idx < 0) return null
    this.treatments[idx] = { ...this.treatments[idx], ...patch, id }
    return this.treatments[idx]
  }

  deleteTreatment(id: string) {
    const idx = this.treatments.findIndex((t) => t.id === id)
    if (idx < 0) return null
    const [removed] = this.treatments.splice(idx, 1)
    return removed
  }

  appointmentCalendar(from: string, to: string) {
    return this.appointments.filter((a) => {
      const d = a.startAt.slice(0, 10)
      return d >= from && d <= to
    })
  }

  chairs() {
    return [
      { id: 'c1', number: 1, name: 'Chair 1', status: 'available' },
      { id: 'c2', number: 2, name: 'Chair 2', status: 'available' },
      { id: 'c3', number: 3, name: 'Chair 3', status: 'available' },
    ]
  }

  staffSchedules(date?: string) {
    const d = date || todayISO()
    return [
      { id: 'ss1', staffName: 'Dr. Kimura', role: 'dentist', date: d, startTime: '09:00', endTime: '18:00', notes: '' },
      { id: 'ss2', staffName: 'Hygienist Hayashi', role: 'hygienist', date: d, startTime: '09:00', endTime: '17:00', notes: 'PMTC' },
    ]
  }

  noShowAppointments() {
    return this.appointments.filter((a) => a.noShow || a.status === 'no_show')
  }

  patientProfile(id: string) {
    const p = this.patient(id)
    if (!p) return null
    return {
      ...p,
      gender: 'male',
      address: 'Tokyo',
      insurance: {
        insurerName: 'Kenpo',
        symbol: '1234',
        number: '567890',
        copayCategory: '30%',
        validUntil: '2027-03-31',
      },
      medicalHistories: [],
      allergyRecords: [],
      familyMembers: [],
      questionnaires: [],
      emergencyContacts: [],
      visitHistory: [],
    }
  }

  xrayImages(patientId?: string) {
    return this.xrays.filter((x) => !patientId || x.patientId === patientId)
  }

  perioExamsForPatient(patientId: string) {
    return this.perioExams.filter((e) => e.patientId === patientId)
  }

  createPerioExam(input: Omit<PerioExam, 'id' | 'createdAt'>) {
    const record: PerioExam = {
      id: `pe${Date.now()}`,
      createdAt: isoNow(),
      ...input,
      sites: input.sites ?? [],
    }
    this.perioExams.unshift(record)
    return record
  }

  updatePerioExam(id: string, patch: Partial<PerioExam>) {
    const idx = this.perioExams.findIndex((e) => e.id === id)
    if (idx < 0) return null
    this.perioExams[idx] = { ...this.perioExams[idx], ...patch, id }
    return this.perioExams[idx]
  }

  deletePerioExam(id: string) {
    const idx = this.perioExams.findIndex((e) => e.id === id)
    if (idx < 0) return null
    const [removed] = this.perioExams.splice(idx, 1)
    return removed
  }

  paginateAuditLogs(page?: number, pageSize?: number) {
    return slicePage(this.auditLogs, page, pageSize ?? 20)
  }

  session() {
    return {
      user: { id: 'u1', email: 'kimura@sakura-dental.jp', name: 'Kimura' },
      organization: {
        id: 'org_demo',
        name: 'Sakura Dental',
        slug: 'sakura-dental',
        planTier: 'PRO',
        subscriptionStatus: 'ACTIVE',
        chairCount: 3,
        memberCount: 4,
      },
      role: 'OWNER',
    }
  }

  organization() {
    return {
      id: 'org_demo',
      name: 'Sakura Dental',
      slug: 'sakura-dental',
      planTier: 'PRO',
      subscriptionStatus: 'ACTIVE',
      chairCount: 3,
      timezone: 'Asia/Tokyo',
      memberCount: 4,
      createdAt: isoNow(),
    }
  }

  usage() {
    return {
      members: 4,
      membersLimit: 20,
      patients: this.patients.length,
      patientsLimit: 500,
      apiCallsThisMonth: 1200,
      apiCallsLimit: 50000,
    }
  }

  teamMembers() {
    return [
      { id: 'm1', role: 'OWNER', joinedAt: isoNow(), lastActiveAt: isoNow(), user: { id: 'u1', email: 'kimura@sakura-dental.jp', name: 'Kimura' } },
      { id: 'm2', role: 'ADMIN', joinedAt: isoNow(), lastActiveAt: isoNow(), user: { id: 'u2', email: 'hayashi@sakura-dental.jp', name: 'Hayashi' } },
    ]
  }

  subscriptionPlans() {
    return [
      { tier: 'STARTER', name: 'Starter', priceMonthly: 9800, priceYearly: 98000, maxMembers: 5, maxPatients: 100, features: [] },
      { tier: 'PRO', name: 'Pro', priceMonthly: 29800, priceYearly: 298000, maxMembers: 20, maxPatients: 500, features: [] },
    ]
  }

  apiKeys() {
    return [{ id: 'key1', name: 'Production', prefix: 'dc_live_', lastUsedAt: null, createdAt: isoNow() }]
  }
}

let singleton: EmbeddedStore | null = null

export function getEmbeddedStore() {
  if (!singleton) singleton = new EmbeddedStore()
  return singleton
}
